### ISSUE TITLE
Fail to add a User using GitHub auth provider

### ISSUE THREAD

**Issue by abretaud** [rancher/rancher#15142]
**Rancher versions:**
rancher/server or rancher/rancher: 2.0.7
rancher/agent or rancher/rancher-agent: 2.0.7

**Docker version: (`docker version`,`docker info` preferred)**

```
Containers: 42
 Running: 22
 Paused: 0
 Stopped: 20
Images: 34
Server Version: 17.12.1-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 9f9c96235cc97674e935002fc3d78361b696a69e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.5.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 32
Total Memory: 31.23GiB
Name: xxxxxxxx
ID: RF2Z:SSI6:XD72:3HPH:6UU3:WYPG:SWYQ:Z73M:RM7T:GECB:RDGS:VMCA
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**

Centos 7 `3.10.0-693.5.2.el7.x86_64`

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)**

Bare-metal

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)**

single node rancher

**Steps to Reproduce:**
In Settings/Authentication, I only select the GiHub method, then "Restrict access to only Authorized Users and Organizations". When I select an existing GitHub user in the list, I get "Principal not found: Internal Server Error" in the UI and this error in the docker logs:

```
2018/08/20 12:32:02 [ERROR] Github getUserOrgById: GET url https://api.github.com/user/%!F(MISSING)%!F(MISSING)17642511 received error from github, err: Request failed, got status code: 404. Response: {
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3/users/#get-a-single-user"
}
```

Tested on v2.0.6 too and other GitHub users with the same result.
I also tried selecting "Allow any valid Users" => the user can login, but I get the same error as an when trying to grant him permission to access a cluster.

**Comment by cjellick** [comment]
Are you using public github or enterprise github?

**Comment by abretaud** [comment]
Public github

**Comment by abretaud** [comment]
Hum, it seems like the problem is just that the url is wrong in https://github.com/rancher/rancher/blob/master/pkg/auth/providers/github/github_client.go#L411
The [github api doc](https://developer.github.com/v3/users/#get-a-single-user) says it should be `https://api.github.com/users/xxx` instead of `https://api.github.com/user/xxx`
I could make a PR for this, but I see no easy way to test it, and I'm a little confused: did the api change (which means some other api calls will probably fail?) or it's just that this code never worked?

**Comment by superseb** [comment]
Can't reproduce this on master, can you describe the exact steps you are taking to get this? And the HTTP calls from Developer Tools in your browser?

**Comment by abretaud** [comment]
- There is only 1 account at the moment: mine, which is a github public account as it is the only auth method I activated
- I log in with this admin github account
- Go to Security > Authentication
- I arrive on the Github authentication settings page (as it's the only method)
- In the "Site access" block, "Restrict access to only Authorized Users and Organizations" is selected
- In "Authorized Users and Organizations", there is only my account
- In the select box, I begin to enter the login of the user I want to give credentials (@mboudet)
- The user appears in the list, I click on it
- I get the error message: 
![capture du 2018-09-11 14-53-11](https://user-images.githubusercontent.com/238755/45361432-6b22c300-b5d2-11e8-83d9-4bce7993df02.png)

Here's the http call throwing the error (HTTP 500):
method: GET
url: https://foo.bar/v3/principals/github_user%3A%2F%2F17642511
response:
```
{"baseType":"error","code":"ServerError","message":"Request failed, got status code: 404. Response: {\n  \"message\": \"Not Found\",\n  \"documentation_url\": \"https://developer.github.com/v3/users/#get-a-single-user\"\n}","status":500,"type":"error"}
```

If it's of any help, here's the http call that propose suggestions of github users for the select box:
method: POST (payload = `{name: "mbou"}`)
url: https://foo.bar/v3/principals?action=search
response:
```
{"type":"collection","links":{"self":"https://foo.bar/v3/principals"},"actions":{"search":"https://foo.bar/v3/principals?action=search"},"sort":{"reverse":"https://foo.bar/v3/principals?action=search\u0026order=asc","links":{"loginName":"https://foo.bar/v3/principals?action=search\u0026sort=loginName","name":"https://foo.bar/v3/principals?action=search\u0026sort=name","principalType":"https://foo.bar/v3/principals?action=search\u0026sort=principalType","profilePicture":"https://foo.bar/v3/principals?action=search\u0026sort=profilePicture","profileURL":"https://foo.bar/v3/principals?action=search\u0026sort=profileURL","provider":"https://foo.bar/v3/principals?action=search\u0026sort=provider","uuid":"https://foo.bar/v3/principals?action=search\u0026sort=uuid"}},"filters":{"created":null,"creatorId":null,"id":null,"loginName":null,"me":null,"memberOf":null,"name":null,"principalType":null,"profilePicture":null,"profileURL":null,"provider":null,"removed":null,"uuid":null},"resourceType":"principal","data":[{"baseType":"principal","created":null,"creatorId":null,"id":"github_user://17642511","links":{"self":"https://foo.bar/v3/principals/github_user:%2F%2F17642511"},"loginName":"mboudet","me":false,"memberOf":false,"name":"mboudet","principalType":"user","profilePicture":"https://avatars0.githubusercontent.com/u/17642511?v=4","provider":"github","type":"principal"},{"baseType":"principal","created":null,"creatorId":null,"id":"github_user://35401476","links":{"self":"https://foo.bar/v3/principals/github_user:%2F%2F35401476"},"loginName":"mboudette","me":false,"memberOf":false,"name":"mboudette","principalType":"user","profilePicture":"https://avatars1.githubusercontent.com/u/35401476?v=4","provider":"github","type":"principal"}]}
```

**Comment by cjellick** [comment]
Thanks for the info. We'll look into this

**Comment by abretaud** [comment]
Still affected by this bug in 2.1.1. Any news on a possible fix?
I was wrong in https://github.com/rancher/rancher/issues/15142#issuecomment-416936610 : the github url is good, it's just that the [`externalID` passed to `getUserOrgByID`](https://github.com/rancher/rancher/blob/master/pkg/auth/providers/github/github_provider.go#L271) still begins with two slashes. They don't seem to be trimmed as they should at https://github.com/rancher/rancher/blob/master/pkg/auth/providers/github/github_provider.go#L259

**Comment by cjellick** [comment]
We've not been able to reproduce and haven't seen any other reports of it. This issue has me stumped because we cannot reproduce and we haven't had any other reports of it. GitHub is our #1 auth provider, so if it was a widespread problem, I'm sure we would see more reports of this.
Are you using a non-standard browser? What's your browser and version?

**Comment by abretaud** [comment]
I use firefox 63.0 on a fedora 28. Nothing exotic on it, and my colleague have the same problem on our instance.
I begin to wonder if it could be a problem with our proxy doing nasty things with the slashes in github_user://xxxxx when it's passed in the url. I'll investigate it and let you know.

**Comment by abretaud** [comment]
Ok, so I confirm it is a problem with our ssl proxy, sorry for the noise.
It is an apache proxy, configured like this:
```
    ProxyPass / http://docker_host:8080/
    ProxyPassReverse /  http://docker_host:8080/
```
(Rancher runs on the docker_host machine)

Modifying like this help a little:
```
    ProxyPass / http://docker_host:8080/  nocanon
    ProxyPassReverse /  http://docker_host:8080/
```

But I still have problem with this http call:

`GET https://our_rancher_fqdn/v3/principals/github_user://xxxxxxx`

Apache doesn't like double slashes, and sends back a 301 response to redirect to `https://our_rancher_fqdn/v3/principals/github_user:/xxxxxxx` => rancher fails to find the github user (it fails to parse the `github_user:/xxxxxxx` part)

I'm trying to change the apache config to accept the double slashes, but it seems complicated... Any clue on this?

**Comment by abretaud** [comment]
In fact it seems like double slashes in urls are not very standard which explains why apache collapse it into a single slash: https://stackoverflow.com/questions/8699542/why-is-apache-removing-multiple-slashes

Maybe the Rancher code could be changed to avoid using :// (or POST it instead of using it in GET requests)? I wonder if it's widely used in Rancher for other things than authentication

**Comment by abretaud** [comment]
Closing this as my problem is solved. Some explanation in case anyone have similar problem one day:

- the problem was that for some reason (see below) the `github_user://xxxx` part of the url was not encoded before being sent to apache. Which made apache send the 301 redirect to remove the double slash

- looking at [the UI code](https://github.com/rancher/ui/blob/master/lib/shared/addon/components/input-identity/component.js#L64), the parameter seems to be encoded correctly. In fact, I must have had some old broken js code in my browser cache, as suddenly it started to work after flushing it. I tested with another browser and could confirm that it works too.

I don't really understand how I didn't detect this cache problem earlier, that's one of the first thing I try usually. So much time wasted... Sorry for the noise :/

To sum up, I just had to configure our apache https proxy like this to make everything work:
```
    ProxyPass / http://docker_host:8080/  nocanon
    ProxyPassReverse /  http://docker_host:8080/
```
